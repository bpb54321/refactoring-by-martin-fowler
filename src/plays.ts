export enum PlayType {
  Tragedy = "tragedy",
  Comedy = "comedy",
}

interface Play {
  name: string;
  type: PlayType;
}

export interface Plays {
  [key: string]: Play;
}

export const plays: Plays = {
  hamlet: { name: "Hamlet", type: PlayType.Tragedy },
  "as-like": { name: "As You Like It", type: PlayType.Comedy },
  othello: { name: "Othello", type: PlayType.Tragedy },
};
