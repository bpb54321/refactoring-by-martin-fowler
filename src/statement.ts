import type { Invoice, Performance } from "./invoices";
import { type Plays, PlayType } from "./plays";

export function statement(invoice: Invoice, plays: Plays) {
  function getChargeForPerformance(aPerformance: Performance) {
    let result = 0;
    switch (getPlay(aPerformance).type) {
      case PlayType.Tragedy:
        result = 40000;
        if (aPerformance.audience > 30) {
          result += 1000 * (aPerformance.audience - 30);
        }
        break;
      case PlayType.Comedy:
        result = 30000;
        if (aPerformance.audience > 20) {
          result += 10000 + 500 * (aPerformance.audience - 20);
        }
        result += 300 * aPerformance.audience;
        break;
      default:
        throw new Error(`unknown type: ${getPlay(aPerformance).type}`);
    }
    return result;
  }

  function getPlay(performance: Performance) {
    return plays[performance.playID];
  }

  function volumeCreditsFor(aPerformance: Performance): number {
    let result = 0;
    result += Math.max(aPerformance.audience - 30, 0);
    if ("comedy" === getPlay(aPerformance).type) {
      result += Math.floor(aPerformance.audience / 5);
    }
    return result;
  }

  function usd(aNumber: number) {
    return new Intl.NumberFormat("en-US", {
      style: "currency",
      currency: "USD",
      minimumFractionDigits: 2,
    }).format(aNumber / 100);
  }

  function totalVolumeCredits() {
    let volumeCredits = 0;
    for (let perf of invoice.performances) {
      volumeCredits += volumeCreditsFor(perf);
    }
    return volumeCredits;
  }

  let totalAmount = 0;
  let result = `Statement for ${invoice.customer}\n`;

  for (let perf of invoice.performances) {
    // print line for this order
    result += ` ${getPlay(perf).name}: ${usd(getChargeForPerformance(perf))} (${perf.audience} seats)\n`;
  }

  for (let perf of invoice.performances) {
    totalAmount += getChargeForPerformance(perf);
  }

  result += `Amount owed is ${usd(totalAmount)}\n`;
  result += `You earned ${totalVolumeCredits()} credits\n`;
  return result;
}
