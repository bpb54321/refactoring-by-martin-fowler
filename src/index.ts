import { statement } from "./statement";
import { invoices } from "./invoices";
import { plays } from "./plays";

const myStatement = statement(invoices[0], plays);

console.log(myStatement);
