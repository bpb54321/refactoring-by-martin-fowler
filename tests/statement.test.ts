import { invoices } from "../src/invoices";
import { plays, PlayType } from "../src/plays";
import { statement } from "../src/statement";

describe("statement", () => {
  it("should produce the correct statement for invoice 0", () => {
    // act
    const result = statement(invoices[0], plays);

    // assert
    expect(result).toMatchSnapshot();
  });
  it("should throw an error on unknown play type", () => {
    // arrange
    const invoice = {
      customer: "Le théâtre de trente sous",
      performances: [{ playID: "unknownTypePlay", audience: 20 }],
    };
    const plays = {
      unknownTypePlay: {
        name: "Play With Unknown Type",
        type: "my unknown play type" as PlayType,
      },
    };

    // assert
    expect(() => {
      statement(invoice, plays);
    }).toThrowErrorMatchingSnapshot();
  });
});
