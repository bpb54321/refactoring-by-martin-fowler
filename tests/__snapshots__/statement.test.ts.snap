// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statement should produce the correct statement for invoice 0 1`] = `
"Statement for BigCo
 Hamlet: $650.00 (55 seats)
 As You Like It: $580.00 (35 seats)
 Othello: $500.00 (40 seats)
Amount owed is $1,730.00
You earned 47 credits
"
`;

exports[`statement should throw an error on unknown play type 1`] = `"unknown type: my unknown play type"`;
